common/argparse.h

```diff
@@ -0,0 +1,861 @@
+  arg_.meta_action = std::move(action);
+}
+
+template <typename T>
+auto CommandLine::OneOfArgBuilder::OneOfValueT<T>::Default(
+    bool is_default) && -> OneOfValueT {
+  OneOfValueT result = std::move(*this);
+  result.is_default = is_default;
+  return result;
+}
+
+template <typename T>
+CommandLine::OneOfArgBuilder::OneOfValueT<T>::OneOfValueT(llvm::StringRef str,
+                                                          T value)
+    : str(str), value(std::move(value)) {}
+
+template <typename T>
+auto CommandLine::OneOfArgBuilder::OneOfValue(llvm::StringRef str, T value)
+    -> OneOfValueT<T> {
+  return OneOfValueT<T>(str, value);
+}
+
+template <typename T, typename U, size_t N>
+void CommandLine::OneOfArgBuilder::SetOneOf(const OneOfValueT<U> (&values)[N],
+                                            T* result) {
+  static_assert(N > 0, "Must include at least one value.");
+  arg_.is_append = false;
+  OneOfImpl(
+      values, [result](T value) { *result = value; },
+      std::make_index_sequence<N>{});
+}
+
+template <typename T, typename U, size_t N>
+void CommandLine::OneOfArgBuilder::AppendOneOf(
+    const OneOfValueT<U> (&values)[N], llvm::SmallVectorImpl<T>* sequence) {
+  static_assert(N > 0, "Must include at least one value.");
+  arg_.is_append = true;
+  OneOfImpl(
+      values, [sequence](T value) { sequence->push_back(value); },
+      std::make_index_sequence<N>{});
+}
+
+// An implementation tool for the one-of value candidate handling. Delegating to
+// this allows us to deduce a pack of indices from the array of candidates, and
+// then use that variadic pack to operate on the array in the variadic space.
+// This includes packaging the components up separately into our storage
+// representation, as well as processing the array to find and register any
+// default.
+//
+// The representation is especially tricky because we want all of the actual
+// values and even the *type* of values to be erased. We do that by building
+// lambdas that do the type-aware operations and storing those into type-erased
+// function objects.
+template <typename U, size_t N, typename MatchT, size_t... Indices>
+void CommandLine::OneOfArgBuilder::OneOfImpl(
+    const OneOfValueT<U> (&input_values)[N], MatchT match,
+    std::index_sequence<Indices...> /*indices*/) {
+  std::array<llvm::StringRef, N> value_strings = {input_values[Indices].str...};
+  std::array<U, N> values = {input_values[Indices].value...};
+
+  // Directly copy the value strings into a heap-allocated array in the
+  // argument.
+  new (&arg_.value_strings)
+      llvm::OwningArrayRef<llvm::StringRef>(value_strings);
+
+  // And build a type-erased action that maps a specific value string to a value
+  // by index.
+  new (&arg_.value_action) Arg::ValueActionT(
+      [values, match](const Arg& arg, llvm::StringRef value_string) -> bool {
+        for (int i = 0; i < static_cast<int>(N); ++i) {
```

Comment:
Should this switch to llvm::seq for consistency with elsewhere that you've done that?

Reply:
Good call, I had looked at the .cpp file to try to find these, but not the .h file, thanks!